go/tools/gazelle/packages/walk.go

```diff
@@ -52,3 +66,103 @@ func Walk(bctx build.Context, root string, f WalkFunc) error {
 		return f(pkg)
 	})
 }
+
+func findPackage(bctx build.Context, dir string) (*build.Package, error) {
+	packageGoFiles, otherFiles, err := findPackageFiles(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	packageName, err := selectPackageName(packageGoFiles, dir)
+	if err != nil {
+		return nil, err
+	}
+
+	var files []os.FileInfo
+	files = append(files, packageGoFiles[packageName]...)
+	files = append(files, otherFiles...)
+	sort.Slice(files, func(i, j int) bool {
+		return files[i].Name() < files[j].Name()
+	})
+	bctx.ReadDir = func(dir string) ([]os.FileInfo, error) {
+		return files, nil
+	}
+	return bctx.ImportDir(dir, build.ImportComment)
+}
+
+func findPackageFiles(dir string) (packageGoFiles map[string][]os.FileInfo, otherFiles []os.FileInfo, err error) {
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return
+	}
+
+	packageGoFiles = make(map[string][]os.FileInfo)
+	for _, file := range files {
+		if file.IsDir() {
+			continue
+		}
+
+		name := file.Name()
+		filename := filepath.Join(dir, name)
+		ext := path.Ext(name)
+		isGo := ext == ".go"
+
+		if !isGo {
+			otherFiles = append(otherFiles, file)
+			continue
+		}
+		fset := token.NewFileSet()
+		ast, err := parser.ParseFile(fset, filename, nil, parser.PackageClauseOnly)
+		if err != nil {
+			log.Printf("%s: error parsing package clause: %v", filename, err)
+			continue
+		}
+
+		packageName := ast.Name.Name
+		if packageName == "documentation" {
+			// go/build ignores this package.
+			continue
+		}
+		if strings.HasSuffix(packageName, "_test") {
+			packageName = packageName[:len(packageName)-len("_test")]
+		}
+		packageGoFiles[packageName] = append(packageGoFiles[packageName], file)
+	}
+	return
+}
+
+func defaultPackageName(dir string) string {
+	pname := filepath.Base(dir)
+	if pname == "." || pname == "/" {
+		// We'll only use this name at the root of the filesystem.
+		return "unnamed"
+	}
+	return pname
+}
+
+func selectPackageName(packageGoFiles map[string][]os.FileInfo, dir string) (string, error) {
+	if len(packageGoFiles) == 0 {
+		return "", &build.NoGoError{Dir: dir}
+	} else if len(packageGoFiles) == 1 {
+		var packageName string
+		for name, _ := range packageGoFiles {
+			packageName = name
+		}
+		return packageName, nil
+	} else {
+		defaultName := defaultPackageName(dir)
+		if _, ok := packageGoFiles[defaultName]; ok {
+			return defaultName, nil
+		}
+
+		err := &build.MultiplePackageError{Dir: dir}
+		for name, files := range packageGoFiles {
+			// Add the first file for each package for the error message.
+			// Error() method expects these lists to be the same length. File
+			// lists are guaranteed to be non-empty.
+			err.Packages = append(err.Packages, name)
+			err.Files = append(err.Files, files[0].Name())
```

Comment:
I've updated the comment a bit to explain. `findPackageFiles` will only create an entry in `packageGoFiles` when it finds a .go file. So the list should never be empty.